QEfficient/transformers/modeling_spd_utils.py

```diff
@@ -0,0 +1,42 @@
+# -----------------------------------------------------------------------------
+#
+# Copyright (c) 2024 Qualcomm Innovation Center, Inc. All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# -----------------------------------------------------------------------------
+
+from typing import Optional
+
+import torch
+
+def filter_hidden_states(
+    hidden_states: torch.Tensor,
+    position_ids: torch.Tensor,
+    num_logits_to_keep: Optional[int] = None,
+) -> torch.Tensor:
+    """
+    Filter hidden states based on whether this is a TLM SpD model
+
+    ``Mandatory`` Args:
+        :hidden_states (torch.Tensor): Hidden states tensor.
+        :position_ids (torch.Tensor): Position ids tensor.
+    ``Optional`` Args:
+        :num_logits_to_keep (int, optional): Number of speculative tokens, specified only for TLM SpD model
+
+    Returns:
+        :torch.Tensor: Filtered hidden states.
+    """
+    batch_size = position_ids.size(0)
+    batch_indices = torch.arange(batch_size)
+    # Cast to INT32 to avoid issue while running in ONNXRT
+    logit_index = position_ids.to(torch.int32).argmax(1, keepdim=True)
+    if num_logits_to_keep is None:
+        # return the last logit
+        return hidden_states[batch_indices.view(-1, 1), logit_index]
+    # gather approach
+    lower_idx = torch.where(logit_index <= num_logits_to_keep, 0, logit_index - num_logits_to_keep).view(-1,1) # shape: [bsz, 1]
```

Comment:
Shouldn't we just error out when `logit_index>num_logits_to_keep`?

Reply:
### Gather Change
looking at this gather implementation, I actually noticed an error and changed accordingly. Validated this new implementation is "correct" by playing with `num_logits_to_keep` on below code:
```python
import torch

position_ids = torch.tensor([
    [0,-1,-1,-1,-1],
    [0,1,-1,-1,-1],
    [0,1,2,-1,-1],
    [0,1,2,3,-1], 
    [0,1,2,3,4], 
])

num_logits_to_keep = 3
batch_size = position_ids.size(0)
batch_indices = torch.arange(batch_size)
# Cast to INT32 to avoid issue while running in ONNXRT
logit_index = position_ids.to(torch.int32).argmax(1, keepdim=True)
# gather approach
lower_idx = torch.where(logit_index < num_logits_to_keep, 0, logit_index+1 - num_logits_to_keep).view(-1,1) # shape: [bsz, 1]
# lower_idx output
tensor([[0],
        [0],
        [0],
        [1],
        [2]])
```

### `logit_index>num_logits_to_keep`
if `logit_index>num_logits_to_keep`, this means that to compute the `lower_idx` to extract according subset of logits, we can do something like below:
`lower_idx = logit_index+1 - num_logits_to_keep`

as such, why would we throw an error in this case?